src/utils/distributions.cpp

```diff
@@ -18,3 +19,129 @@ double bayesmix::multi_normal_prec_lpdf(const Eigen::VectorXd &datum,
+                                       Eigen::VectorXd weights2) {
+  std::vector<Eigen::MatrixXd> vars1;
+  std::vector<Eigen::MatrixXd> vars2;
+
+  for (const auto &p : precs1) vars1.push_back(stan::math::inverse_spd(p));
+
+  for (const auto &p : precs2) vars2.push_back(stan::math::inverse_spd(p));
+
+  double mix1 = 0.0;
+  for (int i = 0; i < means1.size(); i++) {
+    for (int j = 0; j < means1.size(); j++) {
+      mix1 += weights1(i) * weights1(j) *
+              std::exp(stan::math::multi_normal_lpdf(means1[i], means1[j],
+                                                     vars1[i] + vars1[j]));
+    }
+  }
+
+  double mix2 = 0.0;
+  for (int i = 0; i < means2.size(); i++) {
+    for (int j = 0; j < means2.size(); j++) {
+      mix2 += weights2(i) * weights2(j) *
+              std::exp(stan::math::multi_normal_lpdf(means2[i], means2[j],
+                                                     vars2[i] + vars2[j]));
+    }
+  }
+
+  double inter = 0.0;
+  for (int i = 0; i < means1.size(); i++) {
+    for (int j = 0; j < means2.size(); j++) {
+      inter += weights1(i) * weights2(j) *
+               std::exp(stan::math::multi_normal_lpdf(means1[i], means2[j],
+                                                      vars1[i] + vars2[j]));
+    }
+  }
+
+  return mix1 + mix2 - 2 * inter;
+}
+
+double bayesmix::gaussian_mixture_dist(
+    std::vector<bayesmix::MarginalState::ClusterState> clus1,
+    Eigen::VectorXd weights1,
+    std::vector<bayesmix::MarginalState::ClusterState> clus2,
+    Eigen::VectorXd weights2) {
+  double out;
+
+  if (clus1[0].has_univ_ls_state()) {
+    Eigen::VectorXd means1(clus1.size());
+    Eigen::VectorXd vars1(clus1.size());
+    Eigen::VectorXd means2(clus2.size());
+    Eigen::VectorXd vars2(clus2.size());
+    for (int i = 0; i < clus1.size(); i++) {
+      means1(i) = clus1[i].univ_ls_state().mean();
+      vars1(i) = clus1[i].univ_ls_state().var();
+    }
+
+    for (int i = 0; i < clus2.size(); i++) {
+      means2(i) = clus2[i].univ_ls_state().mean();
+      vars2(i) = clus2[i].univ_ls_state().var();
+    }
+
+    out = gaussian_mixture_dist(means1, vars1, weights1, means2, vars2,
+                                weights2);
+  } else if (clus1[0].has_multi_ls_state()) {
+    std::vector<Eigen::VectorXd> means1, means2;
+    std::vector<Eigen::MatrixXd> precs1, precs2;
+
+    for (const auto &c : clus1) {
+      means1.push_back(bayesmix::to_eigen(c.multi_ls_state().mean()));
+      precs1.push_back(bayesmix::to_eigen(c.multi_ls_state().prec()));
+    }
+
+    for (const auto &c : clus2) {
+      means2.push_back(bayesmix::to_eigen(c.multi_ls_state().mean()));
+      precs2.push_back(bayesmix::to_eigen(c.multi_ls_state().prec()));
+    }
+
+    out = gaussian_mixture_dist(means1, precs1, weights1, means2, precs2,
+                                weights2);
+  } else {
+    throw std::invalid_argument("Parameter type not recognized");
```

Comment:
What's the difference between throwing an std::invalid_argument() and simply calling it?

Reply:
I have no idea :) I saw on the stan/math library that they always use throw and that they are also able of testing whether the error gets triggered and I stuck to it

Reply:
I just checked and the behaviour seems to be the same with or without `throw`. I would remove it just for consistency within the library

Reply:
Ok this is wrong, at least in this case.
If I remove the 'throw' and run the test, not only does the test crash when it should catch the exception, but also the exception message is not printed.
Hence I would check in the rest of the library if you have some dangerous behavior such as this

Reply:
Huh. I will add `throw` everywhere in the library then. I also wanted to make another quick small change, so I'll just open a mini PR after you merge this one